improve tabs performance and UI with this one simple trick

cezaraugusto opened the pull request:
> **dear reviewer each commit addresses a logical part of the whole fix and most of them have tests included. I took 2 days ☕️☕️☕️☕️☕️ rebasing this against the first trial for better review and bug check, so please consider do not squash.**
 
## TL;DR

* Improved performance for opening tabs
* Improved performance for resizing the window
* Tabs do not flicker on hover anymore
* Audio indicator is can now be clicked again to mute/unmute
* Observable tabs now resize automatically as tabs get added/removed/resized
* Going/exiting fullscreen videos gracefully resize tab content
* Fixed back/forward tabPage action not showing the close icon
* Tab Title is always visible once title is fetch
* Tab title can gracefully show emojis :boom: :tada:

# [`Click to ignore it all and go to test plan`](#test-plan)

## Performance tests

### Performance while resizing window
**Test plan was made as the following:**

1. Window is sized as ~1200px
2. 20 tabs loaded with brave.com as webview
3. resizing to minimum then maximizing to 1200px again

Metrics used:

* main resize function cost
* main call stack cost

#### current

##### main resize function cost

We relied on `get tabSize` for our calculations, which has some cost and we don't use it anymore.

<img width="714" alt="before__resize_function_cost" src="https://user-images.githubusercontent.com/4672033/30471381-ec6bbd2a-99ce-11e7-983b-2350834b7dcc.png">

##### main call stack cost

Values varied while doing tests. Below shows that previous call stack was lighter than what this PR introduces so still room for improvement. Getting a tab by id seems to have a heavier cost too and other new methods also added some weight. 

<img width="1219" alt="before__resize_call_stack_size" src="https://user-images.githubusercontent.com/4672033/30471382-ec6e83ac-99ce-11e7-9bf0-8e6a6a637262.png">

#### after this PR

##### main resize function cost

There's no need for this.tabSize getter anymore. 

<img width="682" alt="after__resize_function_cost" src="https://user-images.githubusercontent.com/4672033/30471379-ec61f24a-99ce-11e7-857c-58937abaaa38.png">

##### main call stack cost

<img width="1187" alt="after__resize_call_stack_size" src="https://user-images.githubusercontent.com/4672033/30471378-ec3f04a6-99ce-11e7-84d3-d3690733dccd.png">


### Performance while opening tabs
**Test plan was made as the following:**

1. Window is sized as ~1200px
2. set brave.com as your home page
3. keep one tab only, which should be loaded as brave.com
4. open 19 tabs sequentially, wait all tabs to load (no loading icon). Operation took around 22s in local network

**current**

Again `get tabSize` taking its space which is no longer required.

<img width="676" alt="before__open_tab_function_cost" src="https://user-images.githubusercontent.com/4672033/30471380-ec677ac6-99ce-11e7-90a9-ef3d51180083.png">

<img width="289" alt="summary__tab_open_without_observer" src="https://user-images.githubusercontent.com/4672033/30469896-f49bf79a-99c8-11e7-99f1-32213048b79a.png">

**after this pr**

<img width="280" alt="summary__tab_open_w_observer" src="https://user-images.githubusercontent.com/4672033/30469897-f49f1de4-99c8-11e7-8bde-eb767710dd25.png">

### Performance while closing tabs
**Test plan was made as the following:**

I couldn't see a significant change here, as tab icons are only updated when you mouse out of tabs, but I'm sharing anyway.

1. Window is sized as ~1200px
2. set brave.com as your home page
3. keep one tab only, which should be loaded as brave.com
4. open 20 tabs
4. close 20 tabs sequentially, wait for all tabs to load (no loading icon). Operation took around 22s in local network

Tests were made several times, below a comparison between two consecutive tests.
Other than scripting, no gains were found. Given we only update tab size after mouse leave, that makes sense.

**current**

<img width="291" alt="summary__tab_close_without_observer" src="https://user-images.githubusercontent.com/4672033/30469817-97e48616-99c8-11e7-9eac-6063cb230981.png">

**after this pr**

<img width="285" alt="summary__tab_close_w_observer" src="https://user-images.githubusercontent.com/4672033/30469816-97e09664-99c8-11e7-9ab5-8c5e11851722.png">

### Performance while resizing window -- stress test (multiple heavy tabs)
1. window is sized around 1200px
2. set https://www.youtube.com/watch?v=A_MjCqQoLLA as your homepage (disable your audio pls)
3. Open total of 9 tabs. While profiling operation should took 20s
5. Resize the window to the minimum possible
6. Resize the window back to 1200px

Tests were made several times, below a comparison between two consecutive tests.

* saved around ~400ms from scripting
* saved around ~200ms from rendering
* paiting is about the same for all, which makes sense given icons didn't change

**current**

<img width="297" alt="wo_obs_20s_total_real" src="https://user-images.githubusercontent.com/4672033/30469711-f4bf5a92-99c7-11e7-8fb6-9c1ed1142598.png">

**after this pr**

<img width="284" alt="w_obs_20s_total_real" src="https://user-images.githubusercontent.com/4672033/30469709-f2fac1a6-99c7-11e7-9a4f-08f56db8c838.png">


## Bugs fixed | usability improved

Total: 16 issues

close #10883
fix #10838
fix #10611
fix #10582
fix #10544
fix #10464
fix #10509
fix #10123
fix #10103
fix #9398
fix #8414
fix #7925
fix #7765
fix #7730
fix #7301
fix #6716

## Unrelated tasks covered with this PR
- Opens the road for theming: added theme.js

## <a name="test-plan">Test plan</a>

### For reviewers (QA please skip)
* Most new methods were unit tested, you can track them by looking for each commit description.

### For everybody (cc QA team)

Overall test plan is to have a lot of tabs, being included pinned, partitioned, private, newtab (which has no icon), about:blank (which has a "default" icon) and other tabs. 

As you resize the window, icons should be shown in a graceful state, without bad UI/UX

Running through each issue (from older to newer)

#### Issue #6716 (fading effect b/w tabs feels like flickering)

1. Have 3 tabs open. Ensure at least one have tab color (i.e. facebook), and one is private.
2. Set each tab as active and then hover over other tabs
3. there should be no _flicker_ background (bad UI)

This is slightly different than #10838 as was reported by not the active tab. Test plan is similar.

#### Issue #7301 (Tabs display wrong after exiting HTML5 fullscreen)
1. Have enough tabs so the close tab icon is always visible
2. Also have some pinned tabs
3. In one tab, go to a video website (i.e. youtube)
4. Keep record of what you see in that tab (like favicon, title, close icon and a blue top border)
5. Open a video in fullscreen mode. Keep an eye on tabs just in case
6. Exit fullscreen. 
7. Tab content should be in the same size as step 4


#### Issue #7730 (Title on new tab flickers after setting the language to Russian)

* Set the lang setting to Russian
* Restart
* Open a new tab
* Change the window size vertically
* Title shouldn't ficker

#### Issue #7925 (Throttling between window resize cause tabContent to pop)

1. Open 100 tabs between video playing, new sessions and private tabs. default tabs too
2. resize the window
3. icons shouldn't pop or show/hide with a bad ui feeling

#### Issue #7765 (Tab title often isn't updated until the cursor is moved over it or tab focus is changed)

1. Run simulated Brave payments to generate reports (`npm run add-simulated-payment-history`)
2. Click on the payment report (the "clock" icon) to open it in a new tab
3. Save the file/Close the save file prompt
4. Tab title should show the title and not the url

#### Issue #8414 (Tabs No Longer Show Full-Color Emoji)

1. Open getemoji.com
2. You should see an emoji in the tab title and not a gray/black circle

#### Issue #9398 (Tab Label Rendering Issue)

Not sure how to track this but issue is unreproducible. See #9398 for comments

#### issue #10103 (Tab preview does not work properly with DnD)

**Note:** Needs PR #10937 merged to be tested.

1. open two tabs
2. let your mouse stale for 3s in a tab. this enable preview
3. drag a tab to reorder
4. preview should be canceled

#### Issue #10123 (no tab close icon is shown for the active tab on Windows)

(you need Windows)
(Test plan gently borrowed from @bsclifton)

1. Set tab page to 20
2. Maximize window
3. Create enough tabs that you get a second page (21 tabs)
4. switch back to tab page 1
5. click one of the tabs to make it the active tab
6. Hover over the other tabs and notice there is no X button
7. Close a few tabs- enough that you see the tab breakpoint size change
8. Hover over tabs and X now shows again

#### Issue #10464 (Description for Session Tab on Hover)

1. Have enough tabs so the close tab icon is always visible
2. Open one partitioned tab (new session)
3. Hover over the icon
4. alt title tooltip should render correctly (with a number!)

#### Issue #10509 (make use of intersectionObserver for tabs)

This is the main base of this PR and there's no test plan needed

#### Issue #10544 (Tabs cannot be muted via speaker icon)
1. Open a tab with audio/video playing. See the icon
2. Toggle the icon. Should be muted/unmuted

#### Issue #10582 (Tabs do not automatically resize until mousing over them)

1. Open a lot of tabs up to a point that close icon is always shown
2. Restart the browser
3. important: ensure you do not hover over any tab
4. Resize the window
5. Icons should follow based on window resize

#### Issue #10611 (intermittent - audio indicator is not shown)
This issue happened because when breakpoints was a thing, "dynamic"
breakpoint wasn't set for the icon to be visible (see https://github.com/brave/browser-laptop/blob/master/app/renderer/components/tabs/tab.js#L263). We do not rely on breakpoints anymore, so:

1. Open one tab
2. Go to youtube.com, play a video
3. Audio icon should be visible

#### Issue #10838 (Active tab flicker on hover)
* Have 3 tabs open. Ensure at least one have custom background (i.e. facebook), and one is private.
* Set each tab as active and then hover it
* there should be no _flicker_ background (bad UI)

#### Issue #10883 (create smooth transitions for tab components)
* close icon, favicon, private, new session icons will have a smooth transition between resizes
* audio icon will slightly scroll to the right when enabled

luixxiul commented on less/main.less:
```diff
@@ -3,7 +3,7 @@
 
 @bodyBG: #000;
-@windowContainerFG: #f00;
+@windowContainerFG: #000;
```
main.less is removed on https://github.com/brave/browser-laptop/pull/10861, not being sure what those specifications are required for. See: https://github.com/brave/browser-laptop/pull/10861/files#diff-eb0db025dc73e282655968f53c9bfbdfR171. We could discuss it later.

cezaraugusto replied:
yes I saw that just after pushing. Should be removed here too.

luixxiul replied:
Confirmed main.less was removed here too 👍

NejcZdovc commented on app/common/state/tabContentState/audioState.js:
```diff
@@ -0,0 +1,66 @@
+  const frame = getFrameByKey(state, frameKey)
+
+  if (frame == null) {
+    if (process.env.NODE_ENV !== 'test') {
```
I see that you use this pattern a lot. I am curious why we wouldn't like to see this error in `test` as well?

cezaraugusto replied:
it always falls to that condition for unit tests even when the test pass for some async reason

NejcZdovc replied:
ok will check out unit tests and try to fix it, so that we can remove this check

NejcZdovc commented on app/renderer/components/tabs/content/closeTabIcon.js:
```diff
@@ -50,32 +48,27 @@ class CloseTabIcon extends React.Component {
-      )
-
-    // used in functions
-    props.frameKey = frameKey
```
we need to keep this, because every single ownProps should be passed into new props object

NejcZdovc commented on app/renderer/components/tabs/content/favIcon.js:
```diff
@@ -3,103 +3,127 @@
-    props.isNarrowestView = tabContentState.isNarrowestView(currentWindow, frameKey)
-
-    // used in functions
-    props.frameKey = frameKey
```
ditto

NejcZdovc commented on app/renderer/components/tabs/content/newSessionIcon.js:
```diff
@@ -4,61 +4,55 @@
-      : props.partitionNumber
-
-    // used in functions
-    props.frameKey = frameKey
```
ditto

NejcZdovc commented on app/renderer/components/tabs/content/privateIcon.js:
```diff
@@ -9,49 +9,59 @@ const {StyleSheet, css} = require('aphrodite/no-important')
     props.isActive = frameStateUtil.isFrameKeyActive(currentWindow, frameKey)
-
-    // used in functions
-    props.frameKey = frameKey
```
ditto

NejcZdovc commented on app/renderer/components/tabs/content/tabEndContainer.js:
```diff
@@ -0,0 +1,92 @@
+const Immutable = require('immutable')
+const ReduxComponent = require('../../reduxComponent')
+
+const {StyleSheet, css} = require('aphrodite/no-important')
```
please add // Styles

NejcZdovc commented on app/renderer/components/tabs/content/tabEndContainer.js:
```diff
@@ -0,0 +1,92 @@
+
+const React = require('react')
+const Immutable = require('immutable')
+const ReduxComponent = require('../../reduxComponent')
```
please add // Components

NejcZdovc commented on app/renderer/components/tabs/content/tabEndContainer.js:
```diff
@@ -0,0 +1,92 @@
+const {StyleSheet, css} = require('aphrodite/no-important')
+const globalStyles = require('../../styles/global')
+
+const frameStateUtil = require('../../../../../js/state/frameStateUtil')
```
please add // State

NejcZdovc commented on app/renderer/components/tabs/content/tabEndContainer.js:
```diff
@@ -0,0 +1,92 @@
+  mergeProps (state, ownProps) {
+    const currentWindow = state.get('currentWindow')
+    const frame = frameStateUtil.getFrameByKey(currentWindow, ownProps.frameKey) || Immutable.Map()
+    const frameKey = ownProps.frameKey
```
Move this definition one line higher and then use this when defining `frame`

NejcZdovc commented on app/renderer/components/tabs/content/tabEndContainer.js:
```diff
@@ -0,0 +1,92 @@
+    const tabId = frame.get('tabId', tabState.TAB_ID_NONE)
+
+    const props = {}
+    props.children = ownProps.children
```
you need to pass frameKey in new props

NejcZdovc commented on app/renderer/components/tabs/content/tabTitle.js:
```diff
@@ -9,49 +9,48 @@ const {StyleSheet, css} = require('aphrodite/no-important')
-    props.displayTitle = tabContentState.getDisplayTitle(currentWindow, frameKey)
-
-    // used in functions
-    props.frameKey = frameKey
```
ditto

NejcZdovc commented on app/renderer/components/tabs/content/tabTitle.js:
```diff
@@ -9,49 +9,48 @@ const {StyleSheet, css} = require('aphrodite/no-important')
-    // used in functions
-    props.frameKey = frameKey
-
+    props.isWindows = isWindows()
```
this should be cached at the top as it was before

```
const platformUtil = require('../../../../common/lib/platformUtil')
const isWindows = platformUtil.isWindows()		
const isDarwin = platformUtil.isDarwin()
```

NejcZdovc commented on app/renderer/components/tabs/tab.js:
```diff
@@ -2,66 +2,66 @@
+const React = require('react')
+
+// Styles
+const {StyleSheet, css} = require('aphrodite/no-important')
```
this needs to be moved at the top where we have react and immutable

NejcZdovc commented on app/renderer/components/tabs/tab.js:
```diff
@@ -2,66 +2,66 @@
 const ReduxComponent = require('../reduxComponent')
-const Favicon = require('./content/favIcon')
-const AudioTabIcon = require('./content/audioTabIcon')
+
```
please remove this line

NejcZdovc commented on app/renderer/components/tabs/tab.js:
```diff
@@ -189,34 +189,39 @@ class Tab extends React.Component {
   componentDidUpdate () {
-    this.onUpdateTabSize()
+    // if a tab was just pinned, unobserve its intersection
+    if (this.props.isPinnedTab) {
```
we should check if this value was actually changed

```
componentDidUpdate (prevProps) {
    if (this.props.isPinnedTab && !prevProps.isPinnedTab) {
      this.observer.unobserve(this.tabSentinel)
    }
  }
```

NejcZdovc commented on app/renderer/components/tabs/tab.js:
```diff
@@ -231,57 +236,43 @@ class Tab extends React.Component {
-    props.totalTabs = state.get('tabs').size
-    props.dragData = state.getIn(['dragData', 'type']) === dragTypes.TAB && state.get('dragData')
-    props.hasTabInFullScreen = tabContentState.hasTabInFullScreen(currentWindow)
+    props.frameKey = ownProps.frameKey
```
use variable defined at the top `frameKey`

NejcZdovc commented on app/renderer/components/tabs/tab.js:
```diff
@@ -342,44 +328,129 @@ class Tab extends React.Component {
-            ? <TabTitle frameKey={this.props.frameKey} />
-            : null
-          }
+          <Favicon frameKey={this.props.frameKey} />
```
In this component and all bellow we should be passing tabId into them

NejcZdovc commented on docs/state.md:
```diff
@@ -697,7 +696,8 @@ WindowStore
       previewTabPageIndex: number, // index of the tab being previewed
-      tabPageIndex: number // index of the current tab page
+      tabPageIndex: number, // index of the current tab page
+      intersectionRatio: number // at which tab size position the tab sentinel is being intersected
```
please sort this alphabetically

NejcZdovc commented on js/state/frameStateUtil.js:
```diff
@@ -717,6 +722,38 @@ const setTabHoverState = (state, frameKey, hoverState, enablePreviewMode) => {
+  const tabPageIndex = state.getIn(['ui', 'tabs', 'tabPageIndex'], 0)
+  const previewTabPageIndex = state.getIn(['ui', 'tabs', 'previewTabPageIndex'])
+
+  return previewTabPageIndex != null ? previewTabPageIndex : tabPageIndex
```
can we simplify this to `return state.getIn(['ui', 'tabs', 'previewTabPageIndex']) || tabPageIndex`

NejcZdovc commented on js/state/frameStateUtil.js:
```diff
@@ -717,6 +722,38 @@ const setTabHoverState = (state, frameKey, hoverState, enablePreviewMode) => {
+
+const hasFrame = (state, frameKey) => {
+  const frame = getFrameByKey(state, frameKey)
+  return !frame.isEmpty()
```
`getFrameByKey` can return null as well so we need to add check here as well `frame && !frame.isEmpty()`

NejcZdovc commented on js/state/frameStateUtil.js:
```diff
@@ -717,6 +722,38 @@ const setTabHoverState = (state, frameKey, hoverState, enablePreviewMode) => {
+
+const getTabIdByFrameKey = (state, frameKey) => {
+  const frame = getFrameByKey(state, frameKey)
+  return frame.get('tabId', tabState.TAB_ID_NONE)
```
we need to add null check here as well

NejcZdovc commented on app/renderer/components/tabs/tab.js:
```diff
@@ -189,34 +189,39 @@ class Tab extends React.Component {
+    }
+    // we only have one entry
+    const entry = entries[0]
+    windowActions.setTabIntersectionState(this.props.frameKey, entry.intersectionRatio)
```
This is called quite a lot of time when resizing, can we doubnce it?

cezaraugusto replied:
throttling tabs as we do now leads to #7925. I tried minimizing perf cost by observing only the first tab in a tab set. I'll share perf wins once I finish this monster

cezaraugusto commented on app/common/state/tabUIState.js:
```diff
@@ -0,0 +1,185 @@
+    : defaultColor
+}
+
+module.exports.checkIfTextColorBlackOrWhite = (state, frameKey, color) => {
```
I did that first but it only returns either 'black' or 'white' so I made it more verbose to avoid confusion

NejcZdovc replied:
I think now is a little bit confusion, so if you want to keep this title I would change `color` param to boolean to `isWhite`

cezaraugusto commented on app/common/state/tabContentState/faviconState.js:
```diff
@@ -0,0 +1,88 @@
+    return false
+  }
+
+  const isNewTabPage = frame.get('location') === 'about:newtab'
```
sounds good

cezaraugusto commented on app/common/state/tabContentState/partitionState.js:
```diff
@@ -0,0 +1,49 @@
+    return false
+  }
+
+  return frame.get('partitionNumber')
```
I missed the boolean coercion there `!!frame.get('partitionNumber')` will coerce to falsy (0) to false. nice spot

cezaraugusto commented on app/common/state/tabContentState/partitionState.js:
```diff
@@ -0,0 +1,49 @@
+  return frame.get('partitionNumber')
+}
+
+module.exports.getPartitionNumber = (state, frameKey) => {
```
they're slightly different. I'll keep that one to avoid passing `frame.get('partitionNumber')` as a param. There's also other functions depending on this and all tested. I'll buy the next pancake round if you let this as nit :P

luixxiul commented:
@cezaraugusto just in case let me remind you that detaching and reattaching tabs do not work for now. Also you cannot drag tabs.

NejcZdovc commented:
@luixxiul @cezaraugusto should be fixed with #10937, so I will review it and if it's ok I will merge it very soon

luixxiul commented on app/renderer/components/tabs/content/tabTitle.js:
```diff
@@ -70,28 +62,28 @@ class TabTitle extends React.Component {
 
-  enforceFontVisibility: {
-    fontWeight: '600'
+  tab__title_bold: {
```
I am not sure if this would be a proper name as we have `tab__title_windows` below which has more bold font-weight. How about renaming it to `tab__title_isDarwin`?

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-6c5463a6f60c16525fb939516ad54125R53

On the commit: https://github.com/brave/browser-laptop/commit/d97a56b311bc994f48383dc5c7b0929b0c0bf57a#diff-6c5463a6f60c16525fb939516ad54125R53

luixxiul commented on app/renderer/components/tabs/content/favIcon.js:
```diff
@@ -1,86 +1,87 @@
+            !this.props.showIconAtReducedSize &&
+            css(styles.icon__loading, iconStyles.icon__loading_color)
+          )
+          : !this.props.favicon && css(styles.icon__default, iconStyles.icon__default_props)
```
We have `icon__default*` for both styles and iconStyles. I'm wondering if it would be confusing a bit.

cezaraugusto replied:
changed `icon__default_props` to `icon__default_sizeAndColor`. Not sure if best but more declarative ya

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-af59f73d2e4394ed5cd93fbab7e1cdc7R92

On the commit: https://github.com/brave/browser-laptop/commit/c6947c52b8bbb073cdd257e3e243418e7b4544de#diff-af59f73d2e4394ed5cd93fbab7e1cdc7R92

luixxiul commented on app/renderer/components/styles/theme.js:
```diff
@@ -0,0 +1,59 @@
+      background: '#ddd',
+      borderColor: '#bbb',
+      color: '#000',
+      hover: {
```
would you mind adding lines for readability?

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/blob/83864a8769ee10cb2d2d864b19289ba5e229bd34/app/renderer/components/styles/theme.js

luixxiul commented on app/renderer/components/tabs/content/closeTabIcon.js:
```diff
@@ -49,33 +48,33 @@ class CloseTabIcon extends React.Component {
-      className={css(this.props.showCloseIcon && styles.closeTab)}
+      className={css(
+        styles.closeTab__icon,
+        this.props.centralizeTabIcons && styles.closeTab__icon__centered
```
Shouldn't it be `closeTab__icon_centered`?

luixxiul replied:
Addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-08a3077ca08d52dfaad29de7ae4d904aR76

On the commit: https://github.com/brave/browser-laptop/commit/8a1a150a2777d17b58c52895bf08b7b8a2c33c16#diff-08a3077ca08d52dfaad29de7ae4d904aR76

luixxiul commented on app/renderer/components/tabs/content/audioTabIcon.js:
```diff
@@ -29,71 +28,61 @@ class AudioTabIcon extends React.Component {
     return <TabIcon
-      className={css(styles.icon, styles.icon_audio)}
+      data-test-id={this.audioIcon}
+      className={css(styles.audio__tabIcon)}
```
would you mind renaming this to `styles.audioTa__icon` following https://github.com/brave/browser-laptop/pull/10691/commits/d69c3f7fe2fc37eaee262cf4bcf9fe4bb259e08d#diff-08a3077ca08d52dfaad29de7ae4d904aR75 or renaming the other?

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-1a5be0fc07ffeb6eda18603b57e9fce9R65

On the commit: https://github.com/brave/browser-laptop/commit/01092fe798c23ecd1c6d61c51c3e74e0bfe8d02b#diff-1a5be0fc07ffeb6eda18603b57e9fce9R65

luixxiul commented on app/renderer/components/tabs/content/newSessionIcon.js:
```diff
@@ -26,44 +22,39 @@ const newSessionSvg = require('../../../../extensions/brave/img/tabs/new_session
+
+    const newSessionProps = StyleSheet.create({
+      newSession__indicator: {
+        filter: this.props.isActive && this.props.textIsWhite ? 'invert(100%)' : 'none'
```
Would you mind formatting like this? https://github.com/brave/browser-laptop/commit/f6726227778537061d109964b38f67f19801d4ba#diff-bf09ff10e2191a87584e58393f99b728R44

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-d8880b1566bd03fcd4e0b29425a5d473R50

On the commit: https://github.com/brave/browser-laptop/commit/049f438503ceab704cd9064b4a6811fea1debfda#diff-d8880b1566bd03fcd4e0b29425a5d473R50

luixxiul commented on app/renderer/components/tabs/content/newSessionIcon.js:
```diff
@@ -73,21 +64,17 @@ class NewSessionIcon extends React.Component {
 const styles = StyleSheet.create({
-  icon: {
+  newSession__icon: {
+    zIndex: 99,
```
Would you mind converting this into const and adding it to global.js so that we could see the relationship among z-index values?

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-d8880b1566bd03fcd4e0b29425a5d473R70

On the commit: https://github.com/brave/browser-laptop/commit/049f438503ceab704cd9064b4a6811fea1debfda#diff-d8880b1566bd03fcd4e0b29425a5d473R70

luixxiul commented on app/renderer/components/styles/global.js:
```diff
@@ -214,6 +214,7 @@ const globalStyles = {
     zindexWindowIsPreview: '1100',
     zindexDownloadsBar: '1000',
     zindexTabs: '1000',
+    zindexTabsAudioTopBorder: '10001',
```
Is it not `1001` instead of `10001`?

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-f14732b3ea914a26c2215676e7208907R217

On the commit: https://github.com/brave/browser-laptop/commit/f4780d31478ed4f1091a86f1f19f4c983dd82590#diff-f14732b3ea914a26c2215676e7208907R217

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -301,13 +312,15 @@ class Tab extends React.Component {
+          this.props.isPinnedTab && styles.tab_pinned,
+          this.props.isActive && styles.tab_active,
+          this.props.showAudioTopBorder && styles.tab_audioTopBorder,
+          this.props.isActive && this.props.themeColor && perPageStyles.tab_themeColor,
```
would you mind moving this line up?

luixxiul replied:
@cezaraugusto I'm sorry if I'm wrong, but I thought these should be:

````
          this.props.isActive && styles.active,
          this.props.isActive && this.props.themeColor && perPageStyles.themeColor,
          this.props.isPlayIndicatorBreakpoint && this.props.canPlayAudio && styles.narrowViewPlayIndicator,
````

to group them based on `this.props.isActive`. wdyt?

cezaraugusto replied:
w00t lost in the rebase, let me re-do

cezaraugusto replied:
weird I'm seeing them as you want them to be

<img width="632" alt="screen shot 2017-09-15 at 1 24 37 pm" src="https://user-images.githubusercontent.com/4672033/30493127-61d1ac42-9a19-11e7-9f0d-38b9e98e138c.png">

cezaraugusto replied:
dohh I'm looking for the wrong line

cezaraugusto replied:
in this commit we have:

```js
            styles.tab,
            perPageGradient.tab_gradient,
            // Windows specific style
            isWindows && styles.tab_forWindows,
            this.props.isPinnedTab && styles.tab_pinned,
            this.props.isActive && styles.tab_active,
            this.props.isActive && this.props.themeColor && perPageStyles.tab_themeColor,
            this.props.showAudioTopBorder && styles.tab_audioTopBorder
```

do you think this can be improveed?

luixxiul replied:
LGTM! 👍

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -323,7 +336,10 @@ class Tab extends React.Component {
+        <div className={css(
+          styles.tab__identity,
+          this.props.centralizeTabIcons && styles.tab__content_centered
+          )}>
```
would you please delete two whitespaces before the brackets?

cezaraugusto replied:
wow weird standard accepted this, nice catch

luixxiul replied:
Confirmed this was fixed on https://github.com/brave/browser-laptop/pull/10691/files#diff-4f164dab3a38a1443d7ba3288fef641eR342

On the commit: https://github.com/brave/browser-laptop/commit/94ba1a1557d4646f7e36b69c0f110f59f3b835aa#diff-4f164dab3a38a1443d7ba3288fef641eR331

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -337,34 +353,39 @@ class Tab extends React.Component {
     borderColor: '#bbb',
     boxSizing: 'border-box',
-    color: '#5a5a5a',
+    color: theme.tab.color,
```
On theme.js `theme.tab.color` is defined as `#000`. Is this change expected?

cezaraugusto replied:
you are right

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -323,7 +336,10 @@ class Tab extends React.Component {
-        <div className={css(styles.tabId)}>
+        <div className={css(
+          styles.tab__identity,
+          this.props.centralizeTabIcons && styles.tab__content_centered
```
Shouldn't this be `tab__identify_centered` ?

cezaraugusto replied:
all things that need to be centered are now relying on `tab__content_centered` not only the `tab__identity`. not sure what to comment but lmk what you want to and I'll add

luixxiul replied:
I think this conversation itself should serve as a comment for who are confused ;-) Let's leave it as it is.

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -380,59 +401,56 @@ const styles = StyleSheet.create({
   },
 
-  isPinned: {
+  tab__content_centered: {
```
Ditto.

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -380,59 +401,56 @@ const styles = StyleSheet.create({
-    ':hover': {
-      background: `linear-gradient(to bottom, #fff, ${globalStyles.color.chromePrimary})`
-    }
+  tab_pinned: {
```
would you moving this up under `tab_forWindows` same as https://github.com/brave/browser-laptop/pull/10691/commits/23fcc08cbedecc3a4fbc6a17f56d0a35a23e953b#diff-4f164dab3a38a1443d7ba3288fef641eR315 and https://github.com/brave/browser-laptop/pull/10691/commits/23fcc08cbedecc3a4fbc6a17f56d0a35a23e953b#diff-4f164dab3a38a1443d7ba3288fef641eR316 ?

cezaraugusto replied:
perfect

luixxiul replied:
Confirmed this was addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-4f164dab3a38a1443d7ba3288fef641eR394

On the commit: https://github.com/brave/browser-laptop/commit/f4780d31478ed4f1091a86f1f19f4c983dd82590#diff-4f164dab3a38a1443d7ba3288fef641eR392

luixxiul commented on app/renderer/components/tabs/tab.js:
```diff
@@ -380,59 +401,56 @@ const styles = StyleSheet.create({
 
-  tabId: {
-    justifyContent: 'center',
+  tab_audioTopBorder: {
```
would you mind moving this under `tab_active`?

luixxiul replied:
Addressed on https://github.com/brave/browser-laptop/pull/10691/files#diff-4f164dab3a38a1443d7ba3288fef641eR408

Confirmed on https://github.com/brave/browser-laptop/blame/83864a8769ee10cb2d2d864b19289ba5e229bd34/app/renderer/components/tabs/tab.js#L408

NejcZdovc commented:
@cezaraugusto @luixxiul dnd PR was merged

luixxiul commented:
Currently:

<img width="389" alt="screenshot 2017-09-14 16 51 29" src="https://user-images.githubusercontent.com/3362943/30418175-10745cde-996d-11e7-9d9c-ddb1de72c9ab.png">

`.tab__identity` should have margin-right: 6px to make it into

<img width="400" alt="screenshot 2017-09-14 16 53 28" src="https://user-images.githubusercontent.com/3362943/30418222-41c2ae26-996d-11e7-96ca-b79af4c66947.png">

cezaraugusto commented:
ok PR is ready for review. Thanks @luixxiul and @nejczdovc for the early feedback.

cezaraugusto commented:
cc @nejczdovc I skipped the replacement of frameKey with tabId for this PR. I had several problems with tabs not being found and assertion errors while refactoring helper methods

given the size of this PR I created #10969 to do that as a follow-up

NejcZdovc commented:
@cezaraugusto sure, sounds like a good plan regarding tabId

luixxiul commented:
@cezaraugusto Great work!! 😆 I'm going to go through my comments above and make sure if they were addressed, and then I will do QA on each case on the 1st comment.

luixxiul commented:
I confirmed that most of my comments above were properly addressed, except this one: https://github.com/brave/browser-laptop/pull/10691#discussion_r139171307 @cezaraugusto would you mind checking it out for me? thanks!

luixxiul commented:
Checklist
----

- [ ] Issue #6716 (fading effect b/w tabs feels like flickering)
- [ ] Issue #7301 (Tabs display wrong after exiting HTML5 fullscreen)
- [ ] Issue #7730 (Title on new tab flickers after setting the language to Russian)
- [ ] Issue #7925 (Throttling between window resize cause tabContent to pop)
- [ ] Issue #7765 (Tab title often isn't updated until the cursor is moved over it or tab focus is changed)
- [ ] Issue #8414 (Tabs No Longer Show Full-Color Emoji)
- [ ] Issue #9398 (Tab Label Rendering Issue)
- [ ] issue #10103 (Tab preview does not work properly with DnD)
- [ ] Issue #10123 (no tab close icon is shown for the active tab on Windows)
- [ ] Issue #10544 (Tabs cannot be muted via speaker icon)
- [ ] Issue #10582 (Tabs do not automatically resize until mousing over them)
- [ ] Issue #10611 (intermittent - audio indicator is not shown)
- [ ] Issue #10838 (Active tab flicker on hover)

luixxiul commented:
I'm not quite sure if this fading effect on `com` is not too much.

<img width="276" alt="screenshot 2017-09-16 1 57 34" src="https://user-images.githubusercontent.com/3362943/30494289-7e49d570-9a82-11e7-93a6-d83955939ff5.png">

CC @bradleyrichter for his thought.

luixxiul commented:
Also this transition effect on the pinned tab.

![tab](https://user-images.githubusercontent.com/3362943/30494495-4a593854-9a83-11e7-8105-c39522b345c0.gif)

STR: Open https://mixed-script.badssl.com/, pinned the tab, and hover over it.

luixxiul commented:
No fading effect when you play a video without the speaker icon because of being many tabs opened.

<img width="210" alt="screenshot 2017-09-16 2 06 21" src="https://user-images.githubusercontent.com/3362943/30494592-c29ccef2-9a83-11e7-8fa4-a39cdc13072d.png">

This gif shows the issue better:

![tab](https://user-images.githubusercontent.com/3362943/30494662-0b264784-9a84-11e7-8cee-dc03f413f38f.gif)

luixxiul commented:
During testing the fix for #7625 the browser completely freezes after clicking the link to display the receipt pdf, opening a new window without menu bar.

![tab](https://user-images.githubusercontent.com/3362943/30495769-3c93e44e-9a88-11e7-8d7b-b2bbc9393703.gif)

 
Brave: 0.21.0 
V8: 6.1.534.31 
rev: 83864a8769ee10cb2d2d864b19289ba5e229bd34 
Muon: 4.4.16 
OS Release: 16.7.0 
Update Channel: Release 
OS Architecture: x64 
OS Platform: macOS 
Node.js: 7.9.0 
Brave Sync: v1.3.5 
libchromiumcontent: 61.0.3163.71

luixxiul commented:
Also dragging the tab after minimizing the width of the window displays a part of the next tab:

![tab](https://user-images.githubusercontent.com/3362943/30496451-5eb4e95e-9a8a-11e7-90a0-ee30a078f110.gif)

luixxiul commented:
After tearing the tab, the URL bar on the original window becomes blank until you click the window.

![tab](https://user-images.githubusercontent.com/3362943/30496567-c4371950-9a8a-11e7-859f-45e2c5bd32ef.gif)

luixxiul commented:
tab favicons takes a moment until they are resized and rendered. The lag should be removed I think.

![tab](https://user-images.githubusercontent.com/3362943/30496835-a384fcbc-9a8b-11e7-8355-3b3db9755cb9.gif)

luixxiul commented:
The audio indicator icon looks too light for me.

<img width="43" alt="screenshot 2017-09-16 3 12 02" src="https://user-images.githubusercontent.com/3362943/30497182-df465fce-9a8c-11e7-8287-3b4cfdbb8f08.png">

CC @bradleyrichter

luixxiul commented:
Tested on macOS except #7765 (the prompt is not displayed correctly) and #10123 (this is for Windows). It looks great overall 👍

luixxiul commented:
![clipboard01](https://user-images.githubusercontent.com/3362943/30498087-5c7f2a54-9a90-11e7-9a71-2695317cc559.png)

![clipboard01](https://user-images.githubusercontent.com/3362943/30498132-85164812-9a90-11e7-8f98-19fc44424679.png)

<img width="235" alt="screen shot 2017-08-07 at 11 54 21 pm" src="https://user-images.githubusercontent.com/4672033/29054308-cfd3e1fe-7bcb-11e7-85f4-7c786cf05cdb.png">

cezaraugusto commented:
@luixxiul thanks, nice feedback. so about

* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329839517
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329841013
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329858383

were just personal takes, I'm happy with any decisions we make.

about:
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329842196

I'm not sure the effect you expected. I did a transition effect for the audio icon at first but was removed before the final version of this given too many side effects unrelated to this PR (like double render of tab icon - intermittent)

about: 
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329850074
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329854096
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329854660

mind checking if they're issues in the master branch too? I'm not sure if they are related to this

also, about: 
* https://github.com/brave/browser-laptop/pull/10691#issuecomment-329856323

I'll check what's happening, seems related to transition timing (or the lack of). Let me know if you consider that a blocker so I can label properly and prioritize.

luixxiul commented:
> about: https://github.com/brave/browser-laptop/pull/10691#issuecomment-329842196
>
> I'm not sure the effect you expected.

I expect that the fading effect should be displayed even with the audio bar indicator.

bsclifton commented on app/renderer/components/tabs/content/privateIcon.js:
```diff
@@ -20,18 +23,23 @@ class PrivateIcon extends React.Component {
   mergeProps (state, ownProps) {
     const currentWindow = state.get('currentWindow')
     const frameKey = ownProps.frameKey
+    const hasSeconardImage = tabContentState.hasVisibleSecondaryIcon(currentWindow, ownProps.frameKey)
```
misspelling here; s/hasSeconardImage/hasSecondaryImage/g

cezaraugusto replied:
yep that was just a move of what we had at that moment and method was eventually overridden

bsclifton commented on app/renderer/components/tabs/content/newSessionIcon.js:
```diff
@@ -29,9 +29,11 @@ class NewSessionIcon extends React.Component {
     const frameKey = ownProps.frameKey
     const frame = frameStateUtil.getFrameByKey(currentWindow, frameKey) || Immutable.Map()
     const partition = frame.get('partitionNumber')
+    const hasSeconardImage = tabContentState.hasVisibleSecondaryIcon(currentWindow, ownProps.frameKey)
```
another misspelling- s/hasSeconardImage/hasSecondaryImage/g

cezaraugusto replied:
agreed on but same as https://github.com/brave/browser-laptop/pull/10691#discussion_r139506804

bsclifton commented on app/common/state/tabContentState/privateState.js:
```diff
@@ -0,0 +1,36 @@
+const tabUIState = require('../tabUIState')
+const frameStateUtil = require('../../../../js/state/frameStateUtil')
+
+module.exports.isPrivateTab = (state, frameKey) => {
```
Would it be better to add these methods to `app/common/tabState`?

These methods are getting information about a private tab... but that data would be in any tab.  Also, private tabs never have their state persisted... so having the file looks a little misleading

cezaraugusto replied:
makes sense will move them

bsclifton commented on app/common/state/tabContentState/partitionState.js:
```diff
@@ -0,0 +1,73 @@
+    return 0
+  }
+  const partitionNumber = frame.get('partitionNumber')
+  if (typeof partitionNumber === 'string') {
```
like @NejcZdovc mentioned... I believe the logic of actually parsing the number out exists already in `frameStateUtil.getPartitionNumber()`

cezaraugusto replied:
not really. That method does not accept numbers or number-like arguments. So a partition passing down `1` or `'1'` will return `0`. I did a fiddle to demonstrate: https://jsfiddle.net/cezaraugusto/okrgjtok/

bsclifton commented on app/common/state/tabContentState/partitionState.js:
```diff
@@ -0,0 +1,73 @@
+  const frame = frameStateUtil.getFrameByKey(state, frameKey)
+
+  if (frame == null) {
+    if (process.env.NODE_ENV !== 'test') {
```
There's a lot of these checks... can we consolidate this check into a single method? Maybe it can be added into appConfig (since that already has some environment checks)

bsclifton replied:
If there's a problem with this during unit tests, you should be able to stub `console.error` completely and have it no-op

cezaraugusto replied:
I'm going to remove them all. Looking now they seem silly as frames may be unavailable at start-up time and you'll end up receiving warning unnecessarily which may lead to future bug reports, where in fact there's no bug but only async methods.

bsclifton commented on app/common/state/tabContentState/titleState.js:
```diff
@@ -2,24 +2,64 @@
+
+  if (frame == null) {
+    if (process.env.NODE_ENV !== 'test') {
+      console.error('Unable to find frame for showTabTite method')
```
s/showTabTite/showTabTitle

cezaraugusto replied:
wow eagle eyes. I left a https://github.com/brave/browser-laptop/pull/10691#discussion_r139507556 and TL;DR will eventually remove all warnings

bsclifton commented on js/state/frameStateUtil.js:
```diff
@@ -717,7 +745,21 @@ const setTabHoverState = (state, frameKey, hoverState, enablePreviewMode) => {
   return state
 }
 
+const frameLocationMatch = (frame, location) => {
```
Can we get a unit test for this? 😄

bsclifton replied:
(

bsclifton commented on js/state/frameStateUtil.js:
```diff
@@ -717,7 +745,21 @@ const setTabHoverState = (state, frameKey, hoverState, enablePreviewMode) => {
+    return false
+  }
+
+  return frame.get('location') === location
```
nitpick: we may want to check for `frame.get` before calling it (in case the input is not an immutable.map)

bsclifton commented on app/renderer/components/styles/global.js:
```diff
@@ -25,18 +25,7 @@ const globalStyles = {
-      extraSmall: '40px',
-      smallest: '19px'
-    }
+    breakpointNewPrivateTab: '890px'
```
it might be useful to leave a comment about where this number comes from

bsclifton commented on app/common/state/tabUIState.js:
```diff
@@ -114,3 +117,44 @@ module.exports.centralizeTabIcons = (state, frameKey, isPinned) => {
+  if ((isActive || isHover) && isPrivate) {
+    backgroundColor = theme.tab.active.private.background
+  }
+
```
Because there are multiple checks and assignments for the background color, it would be great to have a unit test for this method to make sure that each case is covered.  With a regex match, it shouldn't be too hard to test (mocking/stubbing may be harder)

For example, if tab is not active and is private, we'd always want to make sure it uses `theme.tab.private.background` (even if isIntersecting is true)

Let me know if you need a hand with this 😄

cezaraugusto commented:
per @luixxiul feedback:

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329839517 (opacity gradient on tab's end) I reduced fading effect my own. Don't want to rush others in this so if looks good for your eyes let's make it and address as feedback comes. Looks decent for my bad eyes. If you have a suggestion this is located under `tabUIState.js`. Last line of that file

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329841013 (transition for pinned tabs background) I removed the gradient for pinned tabs, which made transition slightly better. Let me know if you think is ok. Same as above don't want to rush others into small details and given complexity is low we can rearrange later if that's the case.

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329842196 (top blue border/small sized tab w/ partition removing tab end's gradient) this was an issue with this PR and was addressed, please check out again if you want/have a window. TL;DR: Both gradient and top border were asking for `::before` so I changed one to request for `::after`

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329850074 (browser completely freezes after clicking the link to display the receipt pdf), that's an issue in master branch too and not related to this PR

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329854096 (half sibling tab showing on tab dragging) that was an issue with this PR and was addressed. The observer is larger than some tab's sizes on purpose but native DnD API is buggy. I did a workaround and appreciate if you could give this another round of tests

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329854660 (blank urlbar after tear-off) I couldn't reproduce but bugs in urlbar are very unlikely to be related to changes in here

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329856323 (tab favicons takes a moment until they are resized and rendered), delay happens due to the time the observer is fired. I added a slightly delay for icons to be visible to work around on that. Turns out this addressed another issue (#10883)

* For https://github.com/brave/browser-laptop/pull/10691#issuecomment-329858383 again I did my own and added a more (slightly) darker color. Feel free to jump and change for a btter one if you disagree with decision

cezaraugusto commented:
* For @bsclifon (1st) feedback:

This PR was rebased against mater latest.

All things addressed for https://github.com/brave/browser-laptop/pull/10691#pullrequestreview-63186837

* For early @nejczdovc and @bsclifton feedback
I removed the pattern of making a warning on null frames. It was not accurate anyway given at some start-up time frames would be unavailable for a while and you'll receive a warning message just due to async stuff and bugs report may rise unecessarily.

* For @bscfliton (2nd) feedback - code check:

For https://github.com/brave/browser-laptop/pull/10691/files/ca547ae2cd5a16d83ea737f8264add489e093e78..46fa8208665cce10c5eaa64472accd38fb0a99fc#diff-bf09ff10e2191a87584e58393f99b728 and https://github.com/brave/browser-laptop/pull/10691/files/46fa8208665cce10c5eaa64472accd38fb0a99fc..006abb2cf40564a83c22a04b7bd1bba5d30b5c5f#diff-d8880b1566bd03fcd4e0b29425a5d473 (misspelling word):

Mispelling were eventually removed (see https://github.com/brave/browser-laptop/pull/10691#discussion_r139506804)

* For https://github.com/brave/browser-laptop/pull/10691/files/83864a8769ee10cb2d2d864b19289ba5e229bd34#diff-e8efc1819057d6924042bb83d03a4c40 (add private methods to app/common/tabState):

Per slack discussion not moved. moving things to tabState means making use of tabId which based on local tests not ready for private/partition sessions. Frame props often unreachable leading to misleading tabs (private/partition never shows up)

* For https://github.com/brave/browser-laptop/pull/10691/files/83864a8769ee10cb2d2d864b19289ba5e229bd34#diff-c61d242ccc2bbd364afd55cdb1f32357 (making use of `frameStateUtil.getPartitionNumber()` instead of current new method):

Partition check wasn't changed (pls see https://github.com/brave/browser-laptop/pull/10691#discussion_r139510302)

* For https://github.com/brave/browser-laptop/pull/10691/files/83864a8769ee10cb2d2d864b19289ba5e229bd34#diff-c61d242ccc2bbd364afd55cdb1f32357 (regarding repetitive warning messages) and https://github.com/brave/browser-laptop/pull/10691/files/937c7b74e3b0fbbed1ced37fe83672f329f4ff5a..1a1ea37b58dbaca1b1f5ac3f1992d68ae72e4f43#diff-31e73c2da7af41f8b092d9fc8d945ee4 (proper camelcase for tabTitle method):

All warnings were removed (see https://github.com/brave/browser-laptop/pull/10691#discussion_r139507556)


* For https://github.com/brave/browser-laptop/pull/10691/files/83864a8769ee10cb2d2d864b19289ba5e229bd34#diff-f5a0b45b8323a469e78cf98fb0c49b8f (add tests for frameLocationMatch) and https://github.com/brave/browser-laptop/pull/10691/files/83864a8769ee10cb2d2d864b19289ba5e229bd34#diff-f5a0b45b8323a469e78cf98fb0c49b8f (add check for possible mutable input):

Added test for `frameLocationMatch`. Also added a check for when input is not immutable.

* For https://github.com/brave/browser-laptop/pull/10691/files/049f438503ceab704cd9064b4a6811fea1debfda..e4057591d0a9cff01af5cb17ca422f80b56627f2#diff-f14732b3ea914a26c2215676e7208907 (add comment for `breakpointNewPrivateTab`):

Added a comment. This is not changed with this PR btw but did as per request.

* For https://github.com/brave/browser-laptop/pull/10691/files/94ba1a1557d4646f7e36b69c0f110f59f3b835aa..f4780d31478ed4f1091a86f1f19f4c983dd82590#diff-666abd0d632bb472d8621d0dadae0c8c (add tests for `getTabEndIconBackgroundColor`)

Added tests for getTabEndIconBackgroundColor, they're together with other tabUIState methods.

fyi the method `isFirstFrameKeyInTabPage` was changed too. Instead of mapping through every frame to find the first, added `.first()` for the look-up in the state. more perf wins :)

cezaraugusto commented:
cc @luixxiul tagged you as auditor for #10883 which is added in this PR as well. This solved the unwanted transition after tab page changes, which I think was a nice side effect but agreed that should not be there. Please jump in if you think transitions can be improved. Thanks!

luixxiul commented:
Thanks for addressing. Here is another issue with the top blue border:

<img width="112" alt="screenshot 2017-09-21 12 49 26" src="https://user-images.githubusercontent.com/3362943/30678176-5f1a838c-9ecb-11e7-9b6a-e47282237de3.png">

It would be nice if you could exclude the fading effect from that border. If it requires some other work, I think it is ok to address that as a follow-up. Also for me the speaker un/mute icon does not respond to clicks.

And sometimes while dragging the tabs, their titles are gone as follows:

![tab](https://user-images.githubusercontent.com/3362943/30678315-47ccb4ec-9ecc-11e7-8707-6069c65984d6.gif)

cezaraugusto commented:
thanks @luixxiul I did a commit to address those. TL;DR bad z-index icon and actually there's no need to extend to 100% the gradient so I reduced it. Drag n' drop thing was new to me but addressed too.

I did a new commit for that but will eventually be squashed together with a25ca8f4b6ec3eeec07c0bd1cd6779d15acccf9f

luixxiul commented:
@cezaraugusto thanks for quick fix! I confirmed it should work. 

Yet I get another issue around the tab close button:

<img width="104" alt="screenshot 2017-09-21 14 49 10" src="https://user-images.githubusercontent.com/3362943/30680722-2d01ee10-9edc-11e7-945b-bd10efa05f3c.png">

The right gap between the close button and the edge of the tab gets dark. Is it possible to remove that effect? thanks!

cezaraugusto commented:
unable to remove that's the tab's gradient. I tried to win with paint cost by reducing its size but this made the color more solid per your comment so I added again to what it was. Did a `push -f` on top. Other issues remain solved this is just visual.

cezaraugusto commented on app/renderer/components/tabs/tab.js:
```diff
@@ -414,7 +415,8 @@ const styles = StyleSheet.create({
   tab_audioTopBorder: {
-    '::after': {
+    '::before': {
+      zIndex: globalStyles.zindex.zindexTabsAudioTopBorder,
```
@luixxiul here's our hero for the top border gradient you mentioned

bsclifton commented:
@luixxiul can you please re-review and either approve or deny?  If things look good, we can merge 😄

luixxiul commented:
@cezaraugusto please do not forget to merge this to `0.20.x` too, thanks!

cezaraugusto commented:
master 9e6642f438305ceff0251667ed49f4d71d6eb7f0
0.20.x 472f152dce2174cdad4a944ea231115e4339ae17

luixxiul commented:
By doing QA I must say the tabs area has been massively improved 🦁 Kudos to @cezaraugusto 🎖
